util/osuser/user.go

```diff
@@ -0,0 +1,115 @@
+	"time"
+	"unicode/utf8"
+
+	"tailscale.com/version/distro"
+)
+
+// LookupId is like os/user.LookupId but handles a few edge cases like gokrazy
+// and non-cgo lookups, and returns the user shell.
+func LookupId(uid string) (u *user.User, shell string, err error) {
+	return lookup(uid, user.LookupId)
+}
+
+// Lookup is like os/user.Lookup but handles a few edge cases like gokrazy and
+// non-cgo lookups, and returns the user shell.
+func Lookup(username string) (u *user.User, shell string, err error) {
+	return lookup(username, user.Lookup)
+}
+
+func lookup(usernameOrUID string, lookupStd func(string) (*user.User, error)) (*user.User, string, error) {
+	if runtime.GOOS != "linux" {
+		u, err := lookupStd(usernameOrUID)
+		return u, fallbackShell(), err
+	}
+
+	// No getent on Gokrazy. So hard-code the login shell.
+	if distro.Get() == distro.Gokrazy {
+		shell := "/tmp/serial-busybox/ash"
+		u, err := lookupStd(usernameOrUID)
+		if err != nil {
+			return &user.User{
+				Uid:      "0",
+				Gid:      "0",
+				Username: "root",
+				Name:     "Gokrazy",
+				HomeDir:  "/",
+			}, shell, nil
+		}
+		return u, shell, nil
+	}
+
+	// On Linux, default to using "getent" to look up users so that
+	// even with static tailscaled binaries without cgo (as we distribute),
+	// we can still look up PAM/NSS users which the standard library's
+	// os/user without cgo won't get (because of no libc hooks).
+	// But if "getent" fails, userLookupGetent falls back to the standard
+	// library anyway.
+	return userLookupGetent(usernameOrUID, lookupStd)
+}
+
+func validUsername(usernameOrUID string) bool {
+	maxUid := 32
+	if runtime.GOOS == "linux" {
+		maxUid = 256
+	}
+	if len(usernameOrUID) > maxUid || len(usernameOrUID) == 0 {
+		return false
+	}
+	for _, r := range usernameOrUID {
+		if r < ' ' || r == 0x7f || r == utf8.RuneError { // TODO(bradfitz): more?
+			return false
+		}
+	}
+	return true
+}
+
+func userLookupGetent(usernameOrUID string, lookupStd func(string) (*user.User, error)) (*user.User, string, error) {
+	// Do some basic validation before passing this string to "getent", even though
+	// getent should do its own validation.
+	if !validUsername(usernameOrUID) {
+		return nil, "", errors.New("invalid username or UID")
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	out, err := exec.CommandContext(ctx, "getent", "passwd", usernameOrUID).Output()
+	if err != nil {
+		log.Printf("error calling getent for user %q: %v", usernameOrUID, err)
+		u, err := lookupStd(usernameOrUID)
+		return u, fallbackShell(), err
+	}
+	// output is "alice:x:1001:1001:Alice Smith,,,:/home/alice:/bin/bash"
+	f := strings.SplitN(strings.TrimSpace(string(out)), ":", 10)
+	for len(f) < 7 {
+		f = append(f, "")
+	}
+	return &user.User{
+		Username: f[0],
+		Uid:      f[2],
+		Gid:      f[3],
+		Name:     f[4],
+		HomeDir:  f[5],
+	}, f[6], nil
+}
+
+func fallbackShell() string {
+	if s := os.Getenv("SHELL"); s != "" {
```

Comment:
This doesn't seem right. If I'm looking up a different user on the box, my current SHELL shouldn't influence the info returned about some other user.

This is only applicable if the User we're returning matches os/user.Current() (or is non-Windows and has UID of os.Getuid)

Reply:
Yeah, this was taken from ssh code and is probably not right here.